### ISSUE TITLE
Surprising results

### ISSUE THREAD

**Issue by JohanLarsson** [dotnet/BenchmarkDotNet#601]
```cs
public class UnitVectorBenchmarks
{
    [Benchmark(Baseline = true)]
    public double CreateUnitVectorReturnX()
    {
        return new UnitVector(1, 2).X;
    }

    [Benchmark]
    public double Return1()
    {
        return 1;
    }

    public struct UnitVector
    {
        public readonly double X;
        public readonly double Y;

        public UnitVector(double x, double y)
        {
            X = x;
            Y = y;
        }
    }
}
```

Result in:
```
 |                                 Method |       Mean |     Error |    StdDev | Scaled | ScaledSD | Allocated |
 |--------------------------------------- |-----------:|----------:|----------:|-------:|---------:|----------:|
 |                CreateUnitVectorReturnX |  0.0088 ns | 0.0128 ns | 0.0236 ns |      ? |        ? |       0 B |
 |                                Return1 |  0.1331 ns | 0.0262 ns | 0.0644 ns |      ? |        ? |       0 B |
```

Not sure if I'm dumbing something but it does not look right.

**Comment by adamsitnik** [comment]
Hi @JohanLarsson 

Thanks for reporting the problem.

If you use our `[DisassemblyDiagnoser]` you can see that `new UnitVector(1, 2).X;` gets recognized by JIT as a constant.

```
00007ffe`1c444a90 ConsoleApp15.UnitVectorBenchmarks.CreateUnitVectorReturnX()
00007ffe`1c444a90 c4e17b100507000000 vmovsd xmm0,qword ptr [00007ffe`1c444aa0]

Method got inlined
ConsoleApp15.UnitVectorBenchmarks+UnitVector..ctor(Double, Double)
```

```
00007ffe`1c464a90 ConsoleApp15.UnitVectorBenchmarks.Return1()
00007ffe`1c464a90 c4e17b100507000000 vmovsd xmm0,qword ptr [00007ffe`1c464aa0]
```

If you apply `[MethodImpl(MethodImplOptions.NoOptimization)]` attribute to method `CreateUnitVectorReturnX()` the assembly is:

```
00007ffe`1c474a90 ConsoleApp15.UnitVectorBenchmarks.CreateUnitVectorReturnX()
00007ffe`1c474a9a 33c9            xor     ecx,ecx
00007ffe`1c474a9c 488d45f0        lea     rax,[rbp-10h]
00007ffe`1c474aa0 c4e16957d2      vxorpd  xmm2,xmm2,xmm2
00007ffe`1c474aa5 c4e17a7f10      vmovdqu xmmword ptr [rax],xmm2
00007ffe`1c474aaa 488d4df0        lea     rcx,[rbp-10h]
00007ffe`1c474aae c4e17b101521000000 vmovsd xmm2,qword ptr [00007ffe`1c474ad8]
00007ffe`1c474ab7 c4e17b100d20000000 vmovsd xmm1,qword ptr [00007ffe`1c474ae0]
00007ffe`1c474ac0 e843d3ffff      call    00007ffe`1c471e08
00007ffe`1c474ac5 c4e17b1045f0    vmovsd  xmm0,qword ptr [rbp-10h]

00007ffe`1c474b00 ConsoleApp15.UnitVectorBenchmarks+UnitVector..ctor(Double, Double)
00007ffe`1c474b00 c4e17b1109      vmovsd  qword ptr [rcx],xmm1
00007ffe`1c474b05 c4e17b115108    vmovsd  qword ptr [rcx+8],xmm2
00007ffe`1c474b0b c3              ret
```

and the results are something you would expect:

|                  Method |      Mean |     Error |    StdDev | Scaled |
|------------------------ |----------:|----------:|----------:|-------:|
| CreateUnitVectorReturnX | 1.9125 ns | 0.0093 ns | 0.0078 ns |  1.000 |
|                 Return1 | 0.0020 ns | 0.0058 ns | 0.0049 ns |  0.001 |

**Comment by adamsitnik** [comment]
But in that case, we should display `0 ns`. @AndreyAkinshin I have checked the code generated for this benchmark for 0.10.11 and it's correct:

```cs

#define RETURNS_CONSUMABLE

public delegate System.Double IdleDelegate();

private System.Double __Idle() // __ is to avoid possible name conflict
{
      return default(System.Double);
}

private Consumer consumer = new Consumer();

private void IdleMultiAction(long invokeCount)
{
    for (long i = 0; i < invokeCount; i++)
    {
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
        consumer.Consume(idleDelegate());
    }
}
```

@AndreyAkinshin should we treat it as bug (we should display 0)?

**Comment by AndreyAkinshin** [comment]
Yes, we should definitely display 0. When we see something like (Mean = 0.0020 ns; StdDev = 0.0049 ns), it always means zero. However, people don't look at StdDev/Error and some people don't know that it's impossible to get 0.0020ns as a result on modern hardware. Thus, we can replace it in the summary table by zero.
The problem is: how should we detect such situation? Here are some thoughts:

"Real" method duration cannot be positive and less than 1 CPU cycle. We can come up with a heuristic here. E.g., If we get Mean < 1/2 CPU Cycle and Error/StdDev > CPU cycle, we replace the result by 0. Of course, we have to know the duration of a CPU cycle here. We can do it some `RDTSC`-based assembly injection magic (don't like this approach because it's hard to make it completely portable) or we can use some default hardcoded value (e.g. `0.2ns` which corresponds to `5 GHz`). Unfortunately, we can skip some valid `0ns` cases (which have huge mean value because of huge external noise) and show `0ns` for some very quick method which actually takes exactly 1 or 2 CPU cycles).
The heuristic can be improved by an additional look at the whole set of iteration. When we have a `0ns` method, we usually have some `Target` iterations which work faster than `Idle` iterations. Again, it can be a valid situation for a `2CPU-cycle` method in our noisy world, but if we have enough such `Measurement(Target)<Measurement(Idle)` cases (the threshold value can be also hardcoded), we can show `0ns` and it should be ok for most cases.

**Comment by adamsitnik** [comment]
Fixed by #906
